### PR TITLE
`Graph` can build numpyro model for drawing samples

### PR DESCRIPTION
**Pull request by willGraham01**
Works towards #58 |

Important:

- PR makes `numpyro` required by the package.
- PR, for the time being, assumes that we're only using `numpyro.distribution` objects. I haven't changed all the `DistributionFamily` typehints accordingly, since that will cause a lot of PR noise and impact a lot of the tests.

## What Do?

Provided we stick to objects that `numpyro` can handle, we can use it and the graph structure to sequentially construct a model that can then be used in - for example - an MCMC sampler to draw samples for each of the RVs.

Note that by a "model" here, I specifically refer to "a function that takes no arguments, but which defines a number of sites that form a numpyro model". The [introduction to numpyro](https://num.pyro.ai/en/latest/getting_started.html) page makes this slightly more explicit - the "model" created by a `Graph` is the function analogous to the `eight_schools` function defined in the example.

There are multiple layers to this process. Once we have the `Graph`, we know _how to_ construct the model from the nodes, _but_ would require explicit values for the `ParameterNode`s to actually instantiate a model we could sample from. This results in a slightly complex "nested function" structure for the `Graph.model_constructor` method:

- `Graph.model_constructor` itself returns a function, `_model`.
- `_model` takes a number of keyword arguments equal to the number of `ParameterNode`s in the `Graph`. The keyword arguments must share names with the labels of the `ParameterNode`s.
- Invoking `_model(**parameter_values)` returns the "model" as per the above paragraph.

As such, `_model` is effectively now a function of the `ParameterNode`s in the `Graph`, and thus can (hopefully) be used in optimisers. It can, at the very least, be used to draw samples:

(See also `tests/test_graph/test_model_constructor.py` for a working example)
```python
    # The model builder only needs to be created once per Graph.
    constructor = two_normal_graph.model_constructor()
    assert callable(constructor)

    # Now, we should be able to use each set of parameters to create
    # realisations of the model.
    for param_values in collection_of_parameter_values:
        # Now realise the model with the parameter values we have given
        realisation = constructor(**param_values)
        assert callable(realisation)

        # Can now draw samples from our model, for example.
        # run_nuts_mcmc runs an MCMC on the model defined by `realisation`.
        via_model: dict[str, npt.ArrayLike] = run_nuts_mcmc(
            realisation,
            mcmc_kwargs=mcmc_default_options,
        ).get_samples()
```

## Other Notable Changes / Additions

- `DistributionNode`s have been given a `create_model_site` method. Given all their dependencies, (values for the `ParameterNode`s they depend on, and the `numpyro.sample` object for any `DistributionNode`s they depend on), this returns a `numpyro.sample` of the given `DistributionNode`.
- `ParameterNode`s - I believe - will no longer require the `current_value` attribute, since we effectively turn this into an input argument to the `_model` function we obtain from `Graph.model_constructor`. I haven't implemented this change, because again, scope drift.
- `Graph` has gained a `ordered_dist_nodes` property which is just a convenience wrapper for returning the `DistributionNode`s in dependency order, so the `model_constructor` knows which order to build them in.
- Tests have been added for the `DistributionNode.create_model_site` and `Graph.model_constructor` methods. Note that there's some duplication in the tests, but this is mainly because I couldn't find a non-scope-drifting way to get around this duplication. Have opened #62 to track this, and mark it for fixing.

## Considerations

Some things I considered and would like a 2nd opinion on:

- The output of `Graph.model_constructor` has a case for being an instance of a custom class. 
  - Conceptually, this is because after invoking the `model_constructor`, we have left the world of the graph and now have our model to start performing analysis on.
  - It also means that we'll potentially avoid memory-sharing issues down the line (since `model_constructor` has to make references to the `Graph` attributes, which _might_ change if a user wants to recycle instances). 
  - And finally, it would be really helpful for testing and debugging purposes :sweat_smile: As a class instance, we could have easy access to attributes that we currently _don't_ have access to since they're defined as local variables within functions (within functions).
- Have left a bunch of TODOs in the test files in particular. Again, this is mostly so I don't bloat this PR. #62 was opened with a view to fixing these issues in separate PRs. But this does mean the linter is unhappy (`ruff` is a cruel master).

### CODE REVIEW

File: tests/test_node/test_create_model_site.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+@pytest.fixture
+def run_nuts_mcmc(
+    rng_key: jax.Array,
+) -> MCMCRunner:
+    def inner(model, *, nuts_kwargs=None, mcmc_kwargs=None) -> MCMC:
+        if not nuts_kwargs:
+            nuts_kwargs = {}
+        if not mcmc_kwargs:
+            mcmc_kwargs = {}
```
[the remark this answers is not in the document]
Comment (willGraham01) [reply tests/test_node/test_create_model_site.py:31]: I tried this originally - Ruff still complains (default arguments shouldn't be mutable types) if I do this :cry:

File: tests/test_node/test_create_model_site.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    [
+        pytest.param(
+            DistributionNode(
+                Normal,
+                label="Normal",
+                constant_parameters={"loc": 10.0, "scale": 10.0},
+            ),
+            # This is not a useless lambda, we need to delay evaluation until inside the
+            # model function.
+            lambda: {},  # noqa: PIE807
```
[the remark this answers is not in the document]
Comment (willGraham01) [reply tests/test_node/test_create_model_site.py:52]: I think PIE807 is one of those things that is generally useful, but is going to be a right pain in the tests going forward (since we need to pass "mocks" of parts of the sampling process in to the test arguments, but only want to evaluate them inside the test).

Have noted it in #19 as one to be ignored in the tests. I think in the main codebase it should stay.

File: tests/test_node/test_create_model_site.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (willGraham01) [reply tests/test_node/test_create_model_site.py:31]: We should have a "classic ruff" jar (though I guess #19 suffices...)
